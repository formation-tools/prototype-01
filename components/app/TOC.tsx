import { FNestedTableOfContents, FNestedTableOfContentsEntry } from 'core/types'
import { renderObject } from 'core/renderer'
import Tag, { todoKeywordColorClasses } from 'components/doc/Tag'

import { Disclosure, Transition } from '@headlessui/react'
import { ChevronUpIcon } from '@heroicons/react/20/solid'

type TOCHeading = {
  heading: string
  children?: Array<TOCHeading>
}

export type TOCProps = {
  headings: FNestedTableOfContents
}

// TODO: Use Disclosure to support folding
type TableOfContentsEntryProps = {
  entry: FNestedTableOfContentsEntry
  depth: number
}

function TableOfContentsEntry({
  entry: { heading, children, text },
  depth,
}: TableOfContentsEntryProps) {
  const [transMotion, transOpened, transClosed] = [
    'transition transition-[max-height] duration-300 ease-in-out delay-100',
    'transform max-h-screen',
    'transform max-h-0',
  ]

  const { todoKeyword } = heading

  return (
    <Disclosure as="li" defaultOpen className="py-1 max-w-prose">
      {({ open }) => (
        <>
          <div className="flex items-center gap-1">
            {todoKeyword && (
              <Tag
                content={todoKeyword}
                color={todoKeywordColorClasses(todoKeyword)}
                size="small"
                shape="block"
              />
            )}
            {/* TODO: Implement when headline linking works */}
            <span className="hover:text-blue-700">
              {text.flatMap(renderObject)}
            </span>
            <Disclosure.Button
              as="span"
              className={`${!children.length && 'hidden'} contents`}
            >
              <ChevronUpIcon
                className={[
                  'transition-all',
                  open ? '' : 'rotate-180 transform',
                  'h-5 w-5 cursor-pointer hover:text-blue-700 select-none',
                ].join(' ')}
              />
            </Disclosure.Button>
          </div>
          {children && children.length > 0 && (
            <Transition
              className="overflow-none"
              show={open}
              enter={transMotion}
              enterFrom={transClosed}
              enterTo={transOpened}
              leave={transMotion}
              leaveFrom={transOpened}
              leaveTo={transClosed}
            >
              <Disclosure.Panel
                static
                as="ul"
                className={`px-${2 * depth} h-fit`}
              >
                {children.map((heading, idx) => (
                  <TableOfContentsEntry
                    key={idx}
                    entry={heading}
                    depth={depth + 1}
                  />
                ))}
              </Disclosure.Panel>
            </Transition>
          )}
        </>
      )}
    </Disclosure>
  )
}

export default function TOC({ headings }: TOCProps) {
  if (!headings.length) return null
  return (
    <ul className="text-sm text-gray-600">
      {headings.map((heading, idx) => (
        <TableOfContentsEntry key={idx} entry={heading} depth={1} />
      ))}
    </ul>
  )
}
