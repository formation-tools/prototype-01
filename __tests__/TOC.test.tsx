import { expect, describe, it } from 'vitest'
import { render, screen, waitFor } from '@testing-library/react'

import parse, { extractNestedHeadlines } from 'core/parser'

import TOC from 'components/app/TOC'

const samples = {
  another: [
    '#+TITLE: Test Document',
    '* another one',
    '** another two',
    '*** another three',
    '**** another four',
  ].join('\n'),
  parts: [
    '* Part 1',
    '** Part 1.1 with a very long title that requires space',
    '** Part 1.2',
    '*** Part 1.2.1',
    '* Part 2',
    '* Part 3',
  ].join('\n'),
}

describe('toc', () => {
  const getHeadlines = (text, depth?) =>
    extractNestedHeadlines(parse(text).content, depth)
  it('renders', () => {
    const { asFragment } = render(
      <TOC headings={getHeadlines(samples.parts, 3)} />,
    )
    expect(asFragment()).toMatchSnapshot()
  })
  it('renders maximally until depth 3', async () => {
    render(<TOC headings={getHeadlines(samples.another, 3)} />)
    expect(screen.queryByText(/another one/i)).not.toBe(null)
    expect(screen.queryByText(/another two/i)).not.toBe(null)
    expect(screen.queryByText(/another three/i)).not.toBe(null)
    expect(screen.queryByText(/another four/i)).toBe(null)
  })
})
