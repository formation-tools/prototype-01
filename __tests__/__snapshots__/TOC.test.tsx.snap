// Vitest Snapshot v1

exports[`toc > renders 1`] = `
<DocumentFragment>
  <ul
    class="text-sm text-gray-600 hover:text-black"
  >
    <li
      class="py-1 max-w-prose"
    >
      <div
        class="transition duration-0 hover:duration-150 hover:bg-gray-100 text-gray-600 hover:text-black rounded-md ring-inset hover:ring-1 hover:ring-gray-300 py-1 px-1 flex items-center gap-3"
      >
        <span
          class="py-1 hover:text-blue-700"
        >
          Part 1
        </span>
        <span
          aria-controls="headlessui-disclosure-panel-:r1:"
          aria-expanded="true"
          class="min-w-2 mr-2"
          id="headlessui-disclosure-button-:r0:"
        >
          <svg
            aria-hidden="true"
            class="transition-all  h-5 w-5 cursor-pointer hover:text-blue-700 select-none"
            fill="currentColor"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M14.77 12.79a.75.75 0 01-1.06-.02L10 8.832 6.29 12.77a.75.75 0 11-1.08-1.04l4.25-4.5a.75.75 0 011.08 0l4.25 4.5a.75.75 0 01-.02 1.06z"
              fill-rule="evenodd"
            />
          </svg>
        </span>
      </div>
      <div
        class="overflow-hidden ml-2"
      >
        <ul
          class="ml-2 h-fit"
          id="headlessui-disclosure-panel-:r1:"
        >
          <li
            class="py-1 max-w-prose"
          >
            <div
              class="transition duration-0 hover:duration-150 hover:bg-gray-100 text-gray-600 hover:text-black rounded-md ring-inset hover:ring-1 hover:ring-gray-300 py-1 px-1 flex items-center gap-3"
            >
              <span
                class="py-1 hover:text-blue-700"
              >
                Part 1.1 with a very long title that requires space
              </span>
            </div>
          </li>
          <li
            class="py-1 max-w-prose"
          >
            <div
              class="transition duration-0 hover:duration-150 hover:bg-gray-100 text-gray-600 hover:text-black rounded-md ring-inset hover:ring-1 hover:ring-gray-300 py-1 px-1 flex items-center gap-3"
            >
              <span
                class="py-1 hover:text-blue-700"
              >
                Part 1.2
              </span>
              <span
                aria-controls="headlessui-disclosure-panel-:r6:"
                aria-expanded="true"
                class="min-w-2 mr-2"
                id="headlessui-disclosure-button-:r5:"
              >
                <svg
                  aria-hidden="true"
                  class="transition-all  h-5 w-5 cursor-pointer hover:text-blue-700 select-none"
                  fill="currentColor"
                  viewBox="0 0 20 20"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    clip-rule="evenodd"
                    d="M14.77 12.79a.75.75 0 01-1.06-.02L10 8.832 6.29 12.77a.75.75 0 11-1.08-1.04l4.25-4.5a.75.75 0 011.08 0l4.25 4.5a.75.75 0 01-.02 1.06z"
                    fill-rule="evenodd"
                  />
                </svg>
              </span>
            </div>
            <div
              class="overflow-hidden ml-2"
            >
              <ul
                class="ml-4 h-fit"
                id="headlessui-disclosure-panel-:r6:"
              >
                <li
                  class="py-1 max-w-prose"
                >
                  <div
                    class="transition duration-0 hover:duration-150 hover:bg-gray-100 text-gray-600 hover:text-black rounded-md ring-inset hover:ring-1 hover:ring-gray-300 py-1 px-1 flex items-center gap-3"
                  >
                    <span
                      class="py-1 hover:text-blue-700"
                    >
                      Part 1.2.1
                    </span>
                  </div>
                </li>
              </ul>
            </div>
          </li>
        </ul>
      </div>
    </li>
    <li
      class="py-1 max-w-prose"
    >
      <div
        class="transition duration-0 hover:duration-150 hover:bg-gray-100 text-gray-600 hover:text-black rounded-md ring-inset hover:ring-1 hover:ring-gray-300 py-1 px-1 flex items-center gap-3"
      >
        <span
          class="py-1 hover:text-blue-700"
        >
          Part 2
        </span>
      </div>
    </li>
    <li
      class="py-1 max-w-prose"
    >
      <div
        class="transition duration-0 hover:duration-150 hover:bg-gray-100 text-gray-600 hover:text-black rounded-md ring-inset hover:ring-1 hover:ring-gray-300 py-1 px-1 flex items-center gap-3"
      >
        <span
          class="py-1 hover:text-blue-700"
        >
          Part 3
        </span>
      </div>
    </li>
  </ul>
</DocumentFragment>
`;
